### DISCUSSION TITLE
Access Expired status while keeping ThrowsOnError = true

### DISCUSSION THREAD

**Discussion by Mrizzi-96** [DevTrevi/DgcReader#88] [Q&A] [answered]
I have a question regarding the ThrowsOnError parameter used in `Verify` methods.
**Context**
In a web application, I'd like to catch input exceptions (i.e. signature expired, malformed DGC, etc.) and applicative exceptions (i.e. unavailable rules, refresh errors, etc.), while validation failures should be returned as `DgcValidationStatus`.
**The problem** 
With `ThrowsOnError = false`, I'm not getting any input or applicative exceptions.
With `ThrowsOnError = true`, I'm also getting the exceptions for failed validation (i.e. blacklisted DGC or expired DGC), while I'm just expecting a "not valid" or "expired" validation result (`DgcValidationStatus`).
**Question(s)**
Is the above described behaviour intended? 
Is there a way to access `DgcValidationStatus` while getting input/applicative exceptions?
Is there a way to get a `DgcValidationStatus.Expired = true` with `ThrowsOnError = true`?

**Comment by DevTrevi** [comment, marked as answer]
Hi @ Mrizzi-96, yes this is the expected behavior
Usually, exceptions regarding failed updates are not thrown until the files reaches MaxFileAge, in order to allow the validation to be completed if the service is temporarly unavailable for whatever reason.
When throwOnError is true, every mandatory step in validation will throw an exception, because validation at that point is already completed with a failure, or because a technical problem is preventing the execution of the required steps (e.g. values are not present and can not be downloaded, or have past MaxFileAge)

If you want to access the failure details in order to give additional details to the user, you can access the same properties normally included in the result, catching specific exception types.
For accessing the rules validation status, you should catch `DgcRulesValidationException` or try to cast `DgcException` as `DgcRulesValidationException`:

``` csharp
try
{
    var result = await dgcReader.VerifyForItaly(qrCode, mode, throwOnError: true);
}
catch (DgcSignatureValidationException e)
{
    var signatureResult = e.Result; // This is of type SignatureValidationResult
    // signatureResult == result.Signature

    // e.g:
    Console.WriteLine($"Invalid signature for KID {e.Result.CertificateKid}");
}
catch (DgcBlackListException e)
{
    var blacklistResult = e.Result; // This is of type BlacklistValidationResult
    // blacklistResult == result.Blacklist
}
catch (DgcRulesValidationException e)
{
    var validationResult = e.ValidationResult; // This is of type IRulesValidationResult
    // validationResult == result.RulesValidation

    var status = e.Status;  // <-- HERE YOU CAN ACCESS THE VALIDATION STATUS

    // More specific:
    var italianResult = validationResult.AsItalianValidationResult();
    if (italianResult != null)
    {
        Console.WriteLine(italianResult.ItalianStatus);
        Console.WriteLine(italianResult.StatusMessage);
    }
}
catch (DgcException e)
{
    // Every exception thrown by DgcReader is of type DgcException, even the previous one inherits from this base type.
    // So if you use a global error handler, you can catch DgcException to intercept validation errors thrown by the library
    Console.WriteLine(e);
}

```

**Reply by Mrizzi-96** [reply]
@DevTrevi,
I thank you deeply for your detailed answer. I will do as you suggested!
